### ISSUE TITLE
Shrek 2: less overall lag than real-console

### ISSUE THREAD

**Issue by RetroEdit** [nba-emu/NanoBoyAdvance#312]
Using a TAS of Shrek 2 played back on emulator vs. real console for comparison, I found mGBA is currently [~17.9 seconds faster than real-console](https://github.com/mgba-emu/mgba/issues/2969), while NanoBoyAdvance e170695f9b3283de9145fba40a419e3d06c976e7 is net ~79 VBLANK frames faster than real-console (NBA 1.7.1 is ~81 frames faster than real console because it's an additional 2 frames faster in one level).

For testing, I modified the game to display frame count + "input" count in the top-right; here's a patch: [Shrek_2_GBA_TASHack_0.4.1.bps.zip](https://github.com/nba-emu/NanoBoyAdvance/files/11994419/Shrek_2_GBA_TASHack_0.4.1.bps.zip). Basically, it overwrites the game's input subroutine to playback hardcoded inputs. To use it, it requires starting with no save file and I hope to rework it to be more detailed, easier to use, and overall better in the next few months.

You can see how long a level took by seeing when the frame count freezes at the end of a level and comparing it to real-console. Here's what my results were:
```
Raw original console results:
1-1 end: 09CF,0924, 1-2 end: 13F5,1324, 1-3 end: 1E5D,1D64, 1-4 end: 2B63,29FA, 2-1 end: 3833,365C, 2-2 end: 4539,4333, 2-3 end: 5562,5315, 2-4 end: 5F54,5CAA, 3-1 end: 7780,746E, 3-2 end: 8B08,8610, 3-3 end: 9A2A,9476, 3-4 end: A47C,9EA0, 3-5 end: B386,AD84, 4-1 end: C657,BFC6, 4-2 end: D53C,CE6F, 4-3 end: DF0E,D819, 4-4 end: EADC,E3BD, 4-5 end: F638,EEEF, 5-2 end: 0A5A,028A, 5-3 end: 1D33,14EA, 5-4 end: 2611,1D1D

NBA comparison (level difference, cumulative difference)
1-1 end: (0, 0), 1-2 end: (0, 0), 1-3 end: (0, 0), 1-4 end: (-2, -2), 2-1 end: (0, -2), 2-2 end: (-1, -3), 2-3 end: (-12, -15), 2-4 end: (-20, -35), 3-1 end: (+15, -20), 3-2 end: (-12, -32), 3-3 end: (-5, -37), 3-4 end: (-3, -40), 3-5 end: (0, -40), 4-1 end: (-11, -51), 4-2 end: (-1, -52), 4-3 end: (0, -52), 4-4 end: (-1, -53), 4-5 end: (-1, -54), 5-2 end: (-5, -59), 5-3 end: (-6, -65), 5-4 end: (-14, -79)
```

Basically, on my AGS-101, it will display `2611,1D1D` by the end of 5-4, while on NBA it will display `25C2,1D1D`

**Comment by fleroviux** [comment]
Thanks for this. I am not sure how to address this yet besides hoping that general accuracy improvements will fix this eventually.
But this definitely helps with keeping track of improvements or regressions in emulation accuracy!

**Comment by alyosha-tas** [comment]
I ran this on my GBPlayer and 2-4 end should be 5F45 rather than 5F54

**Comment by RetroEdit** [comment]
> I ran this on my GBPlayer and 2-4 end should be 5F45 rather than 5F54

Fixed now; that typo fix makes the math for 2-4 and 3-1 deltas make much more sense.

----

@alyosha-tas Additionally, I have some newer patches: my older patch has saving, which complicates testing because flash cartridges won't have proper EEPROM timing. I made some patches that remove it. I also made a patch that disables prefetch.

This `.zip` archive contains 3 patches: [Shrek_2_TASHack_nosave.zip](https://github.com/nba-emu/NanoBoyAdvance/files/12553430/Shrek_2_TASHack_nosave.zip)

- 0.5.3_2-3: No saving. Starts in 2-3. I didn't get around to testing this on real-console, but the per-level results on emulator were extremely similar to 0.5.4, with only some variations in a few levels.
- 0.5.4: No saving. Probably ends with `25B0,1D1D` in 5-4.
  - These are the results I have written down for my console testing:
    - `1-1 end: 099D,0924, 1-2 end: 13C1,1324, 1-3 end: 1E25,1D64, 1-4 end: 2B2C,29FA, 2-1 end: 37F9,365C, 2-2 end: 44FD,4333, 2-3 end: 5524,5315, 2-4 end: 5F05,5CAA, 3-1 end: 773C,746E, 3-2 end: 8AC2,8610, 3-3 end: 99E2,9476, 3-4 end: A42F,9EA0, 3-5 end: B337,AD84, 4-1 end: C606,BFC6, 4-2 end: D4E9,CE6F, 4-3 end: DEB9,D819, 4-4 end: EA85,E3BD, 4-5 end: F5DF,EEEF, 5-2 end: 09FD,028A, 5-3 end: 1CD4,14EA, 5-4 end: 25B0,1D1D`
- 0.6.0: No saving; disables prefetch. Should end with `53D0,1D1D` in 5-4.

I'm not entirely happy with my versioning here, but hopefully that at least gives them distinct labels.

**Comment by alyosha-tas** [comment]
I tried 0.5.4 on hardware, and I get 1-3 end is 1E27 instead of 1E25. I also get 1E27 on emulator. I didn't check all the values but I get the ending value of (25B0, 1D1D) on emulator.

**Comment by RetroEdit** [comment]
I rechecked by video footage, and the `1E25` instead of `1E27` was apparently a typo. I rechecked every other value, and that should be the only typo.

I also rechecked my 0.6.0 video, and the ending value of `53D0,1D1D` should be correct as well.
